File: mondrian-charts/src/chart_to_svg/create_line_path_def.rs
use std::fmt::Write;

/// Creates the SVG path definition for a line path.
///
/// `data` should contain the data points for the line.
/// `x`, `y` are a set of factories for producing X and Y coordinates for the
/// line.
///
/// See also: https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d
pub(super) fn create_line_path_def<P>(
    data: &[P],
    x: impl Fn(&P) -> f64,
    y: impl Fn(&P) -> f64,
) -> String {
    let len = data.len();
    if len == 0 {
        return String::new();
    }

    let start = &data[0];
    let mut path = String::with_capacity(12 * data.len());
    write!(&mut path, "M{x:.1},{y:.1}", x = x(start), y = y(start))
        .expect("Could not format line path");

    // Draw a line along the y0 coordinates.
    for point in data.iter().skip(1) {
        write!(&mut path, "L{x:.1},{y:.1}", x = x(point), y = y(point))
            .expect("Could not format line path");
    }

    path
}

File: fiberplane-charts/src/CoreChart/ChartShape/paths/createLinePathDef.ts
import type { CoordinateFactory } from "./types";
import { toFactory } from "./utils";

/**
 * Set of factories for producing X and Y coordinates for a line path.
 */
export type LinePathCoordinateFactories<P> = {
  x: number | CoordinateFactory<P>;
  y: number | CoordinateFactory<P>;
};

/**
 * Creates the SVG path definition for a line shape.
 *
 * @param data The data points for the line.
 * @param coordinates The factories and/or constants to produce the coordinates
 *                    from each data point.
 *
 * See also: https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d
 */
export function createLinePathDef<P>(
  data: Array<P>,
  coordinates: LinePathCoordinateFactories<P>,
): string {
  const len = data.length;
  if (len === 0) {
    return "";
  }

  const x = toFactory(coordinates.x);
  const y = toFactory(coordinates.y);

  const start = data[0];
  let path = `M${x(start).toFixed(1)},${y(start).toFixed(1)}`;

  // Draw a line along the data points.
  for (let i = 1; i < len; ++i) {
    const next = data[i];
    path += `L${x(next).toFixed(1)},${y(next).toFixed(1)}`;
  }

  return path;
}
